### PR TITLE
added additional methods for appendbuf

### PR DESCRIPTION
**Pull request by rrichardson**

### CODE REVIEW

File: src/appendbuf.rs [hunk 1]
```diff
@@ -248,15 +250,105 @@ impl<'a> AppendBuf<'a> {
   /// ```
   ///
   #[inline]
-  pub fn atomic_slice_pos_from_begin(&self, pos: u32, len: u32) -> Result<AROIobuf, ()> {
+  pub fn atomic_slice_pos_from_begin(&self, pos: u32, len: i64) -> Result<AROIobuf, ()> {
```
Comment (cgaebel) [inline src/appendbuf.rs]: Why not i32?

File: src/appendbuf.rs [hunk 2]
```diff
@@ -248,15 +250,105 @@ impl<'a> AppendBuf<'a> {
...
+  #[inline]
+  pub unsafe fn as_limit_slice<'b>(&'b self) -> &'b [u8] {
+    self.raw.as_limit_slice()
+  }
+
+
+  /// Provides an immutable slice from len bytes from the end of the
+  /// written buffer (before the start of the window)
+  #[inline]
+  pub fn slice_from_end<'b>(&'b self, len: u32) -> &'b [u8] { unsafe {
```
Comment (cgaebel) [inline src/appendbuf.rs]: This should probably have a bounds check.

File: src/appendbuf.rs (hunk 1 above)
Comment (rrichardson) [inline src/appendbuf.rs]: I wanted to support the python slices model, where if you enter a negative number, it starts counting backwards from the end.  That said, I didn't do it quite right, and should probably just have two functions that support negative numbers in both directions.

File: src/appendbuf.rs (hunk 2 above)
Comment (rrichardson) [inline src/appendbuf.rs]: Said everyone ever.

File: src/appendbuf.rs (hunk 1 above)
Comment (cgaebel) [inline src/appendbuf.rs]: Yeah I like two functions better. Also should probably be 32 bits.

**Comment by rrichardson** [comment]
OK.  I am very happy with the new slice scheme.   It works like a union of rust's and python's slices.  There is a _from and a _to, but they also take negative numbers, which count back from the end instead of forward from the beginning.

File: src/appendbuf.rs [hunk 3]
```diff
@@ -105,13 +114,21 @@ impl<'a> AppendBuf<'a> {
...
-  ///
   #[inline]
-  pub fn atomic_slice_from_end(&self, len: u32) -> Result<AROIobuf, ()> {
+  pub fn atomic_slice(&self, from: i32, to: i32) -> Result<AROIobuf, ()> {
     unsafe {
       let mut ret = self.raw.clone_atomic();
-      if len > self.raw.lo() { return Err(()) }
-      let lim = (self.raw.lo() - len,self.raw.lo()) ;
+      let start = if from < 0 {
+          self.raw.lo() - !from as u32
```
Comment (cgaebel) [inline src/appendbuf.rs]: What's with the binary nots? It's probably cool, but I don't get it.

**Comment by rrichardson** [comment]
The compliment of -1 is 0,  -10 is 9, etc. Which turns out to be the exact translation I'm after

File: src/appendbuf.rs (hunk 2 above)
Comment (cgaebel) [inline src/appendbuf.rs]: Still needs bounds check here!

Comment (rrichardson) [inline src/appendbuf.rs]: Which part, exactly?

Comment (rrichardson) [inline src/appendbuf.rs]: NM.  I should just remove this, as I ended up not needing it.  Or if it stays, it should conform to the atomic_slice* functions.

Comment (cgaebel) [inline src/appendbuf.rs]: I'm a fan of removing!

Comment (rrichardson) [inline src/appendbuf.rs]: Done.

**Comment by cgaebel** [comment]
Released as iobuf 5.2.4

**Pull request merged**